crates/steel/src/event.rs

```diff
@@ -0,0 +1,193 @@
+    /// See [`Filter::address`].
+    pub fn address<A: Into<ValueOrArray<Address>>>(mut self, address: A) -> Self {
+        self.filter.address = address.into().into();
+        self
+    }
+
+    /// Sets the 1st indexed topic.
+    pub fn topic1<TO: Into<Topic>>(mut self, topic: TO) -> Self {
+        self.filter.topics[1] = topic.into();
+        self
+    }
+
+    /// Sets the 2nd indexed topic.
+    pub fn topic2<TO: Into<Topic>>(mut self, topic: TO) -> Self {
+        self.filter.topics[2] = topic.into();
+        self
+    }
+
+    /// Sets the 3rd indexed topic.
+    pub fn topic3<TO: Into<Topic>>(mut self, topic: TO) -> Self {
+        self.filter.topics[3] = topic.into();
+        self
+    }
+}
+
+#[cfg(feature = "host")]
+mod host {
+    use super::*;
+    use crate::host::HostEvmEnv;
+    use anyhow::{anyhow, Result};
+    use revm::Database as RevmDatabase;
+    use std::fmt::Display;
+
+    impl<D, H: EvmBlockHeader, C> Event<(), &mut HostEvmEnv<D, H, C>>
+    where
+        D: EvmDatabase + Send + 'static,
+        <D as RevmDatabase>::Error: Display + Send + 'static,
+    {
+        /// Constructor for preflighting an event query for a specific Solidity event.
+        ///
+        /// Initializes the environment for event queries, fetching necessary data via the
+        /// [Provider], and generating a storage proof for any accessed elements using
+        /// [EvmEnv::into_input].
+        ///
+        /// [EvmEnv::into_input]: crate::EvmEnv::into_input
+        /// [EvmEnv]: crate::EvmEnv
+        /// [Provider]: alloy::providers::Provider
+        pub fn preflight<S: SolEvent>(
+            env: &mut HostEvmEnv<D, H, C>,
+        ) -> Event<S, &mut HostEvmEnv<D, H, C>> {
+            Event {
+                filter: event_filter::<S, H>(env.header()),
+                env,
+                phantom: PhantomData,
+            }
+        }
+    }
+
+    impl<S: SolEvent, D, H: EvmBlockHeader, C> Event<S, &mut HostEvmEnv<D, H, C>>
+    where
+        D: EvmDatabase + Send + 'static,
+        <D as RevmDatabase>::Error: Display + Send + 'static,
+    {
+        /// Executes the event query using an [EvmEnv] constructed with [Event::preflight].
+        ///
+        /// This uses [tokio::task::spawn_blocking] to run the blocking revm execution.
+        ///
+        /// [EvmEnv]: crate::EvmEnv
+        pub async fn query(self) -> Result<Vec<Log<S>>> {
+            let logs = self
+                .env
+                .spawn_with_db(move |db| db.logs(self.filter))
+                .await
+                .map_err(|err| anyhow!("querying '{}' failed: {}", S::SIGNATURE, err))?;
+            logs.iter()
+                .map(|log| Ok(S::decode_log(log, false)?))
+                .collect()
+        }
+    }
+}
+
+/// Creates an event filter for a specific Solidity event and block header.
+fn event_filter<S: SolEvent, H: EvmBlockHeader>(header: &Sealed<H>) -> Filter {
+    assert!(!S::ANONYMOUS, "Anonymous events not supported");
```

Comment:
It is not possible to cleanly handle anonymous events with this approach. Since do not have a signature success of the decode_log cannot be guaranteed.